### PR TITLE
Make kubernetes dns addon more configurable

### PR DESCRIPTION
**Pull request by juselius**
###### Motivation for this change

Sometimes there is need to have greater control over the cluster DNS, for example augmenting it with additional service discovery (i.e. Consul), or adding DNS records for internal subdomains. With mode Reconcile, all local changes to the maps are instantly overwritten. Changing the Deployment and ConfigMap to EnsureExists allows the user the change the maps after installation.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by srhb** [comment]
Wouldn't this make any subsequent changes in Nix be ignored silently?
cc @johanot

**Comment by juselius** [comment]
@srhb Probably yes, but that's what you want if you have local changes. It's quite easy to track upstream changes:
1. Save the relevant maps
2. Delete them
3. Let addon-manager recreate them 
4. Merge in local changes

**Comment by srhb** [comment]
I just worry that it's unintuitive from the perspective of NixOS. I'll let @johanot chime in though, since he's actually using Kubernetes in production, and I'm not. :)

**Comment by juselius** [comment]
@srhb I'm running kubernetes in production, and I spent quite a few hours trying to come up with a sensible and stable workaround (like using dnsmasq on the host(s)) which would not require messing with the defaults. Nothing really worked out. So unfortunately this was the only practical solution left in the end.

**Comment by johanot** [comment]
@juselius Isn't your usecase an argument for opting out of the kubernetes.dns module completely, i.e. `services.kubernetes.addons.dns.enable = false` and then deploy the dns service manually or add a custom addon to the addon manager with `EnsureExists`? I'd argue that `Reconcile` is the intuitive choice from a declarative point of view.

**Comment by juselius** [comment]
@johanot Well, yes and no. It's much easier to track nixos when a) all you have to do is delete the deployment and configmap to get the new defaults b) ad you have a patch yaml for local  changes. But I see the problem now. Everybody _without_ local modifications end up in dire straits.  Would a an option e.g. "allowLocalChanges" make the patch more palatable?

**Comment by johanot** [comment]
Well yes - especially if the default behavior is preserved.

I think it would be slightly prettier with e.g. `reconcileMode`, `type = enum [ "Reconcile" "EnsureExists" ];`, `default = "Reconcile"`, but that's just opinion. :)

**Comment by juselius** [comment]
@johanot Good idea! I'll post a new PR.

**Review by johanot — CHANGES_REQUESTED** [review]
The more generic option, I think should have a more generic description as well. Consider something along the lines of:
```
description = ''
  Controls the addon manager reconciliation mode for the DNS addon.
  See: <link
  xlink:href="https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/addon-manager/README.md"/>
'';
```

**Comment by juselius** [comment]
Thanks @johanot!

**Comment by srhb** [comment]
@GrahamcOfBorg eval

**Review by johanot — APPROVED** [review]
LGTM now. Thanks @juselius

**Comment by johanot** [comment]
@juselius Can you please squash the entire changeset down to one commit?

**Comment by juselius** [comment]
@johanot done!

**Comment by johanot** [comment]
Oh.. Sorry I didn't see this before. The commit message should be prefixed with `nixos/kubernetes:`, instead of `kubernetes:`, since it involves changes to the NixOS module, not the package.

File: nixos/modules/services/cluster/kubernetes/dns.nix
Comment (Infinisil) [inline nixos/modules/services/cluster/kubernetes/dns.nix]: Mini-nitpick: Descriptions should end with a `.`

**Comment by danbst** [comment]
I find usecases in PR description more compelling than current option description. For a new guy like me, it is not clear why this once setting is exported as NixOS option, but others are not.

Another problem is that only some of "addons" allow this override. That looks a bit ad-hoc. Maybe note about that in option description as well? Otherwise it would require looking into module code to find out what this is for.

**Comment by juselius** [comment]
@danbst Good suggestion! I have added a line in the description about a use case. My opinion regarding the "ad-hocness" of the override,  is that if you need it in the future, do it in the future. There is little point in adding options for things nobody will ever need, just for symmetry's sake.

@johanot Shall we try to get this merged in, before it goes stale again?

**Comment by johanot** [comment]
@juselius Fine by me, but I don't have merge access :-) 
@srhb Do you have any further comments? If not, will you please assist in merging and back-porting?

**Comment by johanot** [comment]
Thanks @danbst !

**Comment by danbst** [comment]
@juselius as for me PR description is still more clear about what this option is for. But maybe you k8s guys use some other language, so :man_shrugging: problem with me I guess.

**Comment by danbst** [comment]
backported

**Comment by juselius** [comment]
First NixOS PR! Time to celebrate :)

@johanot Oh, I thought you had merge access.

@danbst Thanks! Perhaps you are right regarding the description.  But the PR description is just one of many uses. This is nonetheless not an option a k8s newbie normally would find useful (I believe).

**Pull request merged**
